# Mushrooms

<div style="height: 150px; width: 100%; overflow:hidden; position: relative;">
    <div style="position: absolute; color: white; bottom: 20%; left: 3%; z-index: 10;">
        <p style="">
        AI centered investigation on mushroom data set. Using python
        </p>
    </div>
    <img src="./resources/background.jpg" alt="Background mushrooms" style="transform: translateY(-18%);" />
</div>


> <p style="font-size: .8em; color: gray;">Presented by:</p>
> ARGUELLO Camilo <br>
> EHRLE Steffen

## Contents

1. [Introduction](#introduction)
2. [State of art](#state-of-art)
3. [Mushrooms](#mushrooms)
4. [Analysis](#analysis)
5. [Results](#results)
6. [Conclusions](#conclusions)

## Introduction

<p style="text-align: justify"> Eating wild mushrooms can be a treat; they are one of the few plant foods which contain vitamin D and some other benefits for the human beings [3], however, adverse reactions can occur after eating a wild mushroom. Some of those negative reactions include serious respiratory symptoms, with nauseas and vomiting. As consecuence vital organs could be damaged, which can result in death. [1] Thus, the objective for this research is to analyze whether a mushroom is edible or poisonous based on certain characteristics. </p>

## State of art

**MYTHS**

- Mushrooms which grow in a tree are edible. 
- There are no toxic polypore mushrooms.
- If an animal eat is, it is edible for humans.
- we can get sick by smell poisonous mushrooms.
- we can't touch poisonous mushrooms.
- we can't cook poisonous mushrooms.
- Mushroom with bruises blue - toxic.
- Brightly colors - toxic *Color doesn't determine edibility of a fungi.
- Poisonous mushrooms smell and taste awful.
- Poisonous mushrooms always have bright, flashy colors.
- All mushrooms that grow in woods, meadows, and pastures are safe to eat.
- Poisonous mushrooms have pointy caps.

**TRUTH

- They arrive from mycelium (root system), because they dont have roots


**PROVERB: Chinese.

> All fungi are edible. Some fungi are only edible once.
> Terry Pratchett.

**REF**

It  is  very  difficult  to  distinguish  between  edible and  poisonous  mushrooms.  On  the  basis  of  experience, there  are  some  observations,  which  may  help  to  identify the poisonous  mushrooms: bright color,  pink  spores,  a hot burning  taste  or  acidic  flavor,  growing  on  wooden  pieces in  hidden  places,  difficult  to  break  and  bear  a  cup  like structure  (volva)  at  the  base.  Some  of  the   poisonous mushrooms   when   taken   orally   produce   hallucinations. These include toadstools of the  genera Amanita, Psilocybe and Conocybe (Pinson CW et al., 1999).


## Mushrooms

<div class="alert alert-warning">A little remark! This is not the final version, this is only a reference from the state of art. These texts need to be modified to adapt to the project.</div>

A mushroom, or toadstool, is the fleshy, spore-bearing fruiting body of a fungus, typically produced above ground on soil or on its food source.

The standard for the name “mushroom” is the cultivated white button mushroom, Agaricus bisporus; hence the word “mushroom” is most often applied to those fungi (Basidiomycota, Agaricomycetes) that have a stem (stipe), a cap (pileus), and gills (lamellae, sing. lamella) on the underside of the cap. “Mushroom” also describes a variety of other gilled fungi, with or without stems, therefore the term is used to describe the fleshy fruiting bodies of some Ascomycota. These gills produce microscopic spores that help the fungus spread across the ground or its occupant surface.

<img src="http://4.bp.blogspot.com/_d76A8FQ9xm0/SZomQ1rQyeI/AAAAAAAABFc/wQsySaXNqno/s400/mushroom+diagram.jpg" />

Identifying mushrooms requires a basic understanding of their macroscopic structure. Most are Basidiomycetes and gilled. Their spores, called basidiospores, are produced on the gills and fall in a fine rain of powder from under the caps as a result. At the microscopic level the basidiospores are shot off basidia and then fall between the gills in the dead air space. As a result, for most mushrooms, if the cap is cut off and placed gill-side-down overnight, a powdery impression reflecting the shape of the gills (or pores, or spines, etc.) is formed (when the fruit body is sporulating). The color of the powdery print, called a spore print, is used to help classify mushrooms and can help to identify them. Spore print colors include white (most common), brown, black, purple-brown, pink, yellow, and creamy, but almost never blue, green, or red.

Mushroom Features Glossary

- Cap (Pileus): the expanded, upper part of the mushroom; whose surface is the pileus
- Cup (Volva): a cup-shaped structure at the base of the mushroom. The basal cup is the remnant of the button (the rounded, undeveloped mushroom before the fruiting body appears). Not all mushrooms have a cup.

<img src="./resources/cap_shapes.jfif" width="500px" />

- Gills (Lamellae): a series of radially arranged (from the center) flat surfaces located on the underside of the cap. Spores are made in the gills.
- Mycelial threads: root-like filaments that anchor the mushroom in the soli.
- Ring (Annulus): a skirt-like ring of tissue circling the stem of mature mushrooms. The ring is the remnant of the veil (the veil is the tissue that connects the stem and the cap before the gills are exposed and the fruiting body develops). Not all mushrooms have a ring.

<img src="./resources/gill_spacing.jfif" alt="" />

- Scale: rough patches of tissue on the surface of the cap (scales are remnants of the veil).
- Stalk (or Stem, or Stape): the main support of the mushroom; it is topped by the cap. Not all mushrooms have a stalk (stem) [2].

<img src="./resources/annular_rings.jfif" width="500px"/>

<img src="./resources/stipes_stmes.jfif" width="500px" />

Another feature to consider when identifying mushrooms is whether they bruise or bleed a specific color. Certain mushrooms will change colors when damaged or injured. Cutting into a mushroom and observing any color changes can be very important when trying to determine what it is (ref. [4]).

A universal veil is a temporary membranous tissue that fully envelops immature fruiting bodies of certain gilled mushrooms. The developing Caesar’s mushroom (Amanita caesarea), for example, which may resemble a small white sphere at this point, is protected by this structure. The veil will eventually rupture and disintegrate by the force of the expanding and maturing mushroom, but will usually leave evidence of its former shape with remnants. These remnants include the volva, or cup-like structure at the base of the stipe, and patches or “warts” on top of the cap (ref. [5])

A partial veil (also called an inner veil, to differentiate it from the “outer” veil, or velum[1]) is a temporary structure of tissue found on the fruiting bodies of some basidiomycete fungi, typically agarics. Its role is to isolate and protect the developing spore-producing surface, represented by gills or tubes, found on the lower surface of the cap. A partial veil, in contrast to a universal veil, extends from the stem surface to the cap edge. The partial veil later disintegrates, once the fruiting body has matured and the spores are ready for dispersal. It might then give rise to a stem ring, or fragments attached to the stem or cap edge. In some mushrooms, both a partial veil and a universal veil may be present (ref. [6]).

Based on this research we would like analyze the mushroom to determine whether they are edible or not.

## Analysis

Importing the libraries

In [9]:
from sllab import *
from sklearn import preprocessing as pp

# A utility from https://stackoverflow.com/questions/24458645/label-encoding-across-multiple-columns-in-scikit-learn
# Adapted and changed to Labelbonarizer that fits with the whole dataframe.
class MultiColumnLabelEncoder:
    def __init__(self,columns = None):
        self.columns = columns # array of column names to encode
        self.lb = pp.LabelBinarizer()
        
    def fit(self,X,y=None):
        self.lb.fit(np.concatenate(X.to_numpy(), axis=0))
        return self

    def transform(self,X):
        output = X.copy().astype(object)
        if self.columns is None:
            self.columns = output.columns
        
        for col_id, col in enumerate(self.columns):
            #self.lb.fit(output[col])
            transf = self.lb.transform(output[col])
            #output[col] = transf
            for id, el in enumerate(transf):
                output.iloc[id, col_id] = el
                
        return output

    def fit_transform(self,X,y=None):
        return self.fit(X,y).transform(X)
    

Loading the data

In [10]:
data_names = pd.read_csv("expanded",
                    skiprows=9,
                    skipfooter=1)

In [11]:
data_names.head()

EDIBLE  CONVEX  SMOOTH  WHITE  BRUISES  ALMOND  FREE  CROWDED  NARROW  \
0  EDIBLE  CONVEX  SMOOTH  WHITE  BRUISES  ALMOND  FREE  CROWDED  NARROW   
1  EDIBLE  CONVEX  SMOOTH  WHITE  BRUISES  ALMOND  FREE  CROWDED  NARROW   
2  EDIBLE  CONVEX  SMOOTH  WHITE  BRUISES  ALMOND  FREE  CROWDED  NARROW   
3  EDIBLE  CONVEX  SMOOTH  WHITE  BRUISES  ALMOND  FREE  CROWDED  NARROW   
4  EDIBLE  CONVEX  SMOOTH  WHITE  BRUISES  ALMOND  FREE  CROWDED  NARROW   

  WHITE.1  ... SMOOTH.2 WHITE.2 WHITE.3  PARTIAL WHITE.4  ONE  PENDANT  \
0   WHITE  ...   SMOOTH   WHITE   WHITE  PARTIAL   WHITE  ONE  PENDANT   
1    PINK  ...   SMOOTH   WHITE   WHITE  PARTIAL   WHITE  ONE  PENDANT   
2    PINK  ...   SMOOTH   WHITE   WHITE  PARTIAL   WHITE  ONE  PENDANT   
3   BROWN  ...   SMOOTH   WHITE   WHITE  PARTIAL   WHITE  ONE  PENDANT   
4   BROWN  ...   SMOOTH   WHITE   WHITE  PARTIAL   WHITE  ONE  PENDANT   

   PURPLE  SEVERAL  WOODS  
0   BROWN  SEVERAL  WOODS  
1  PURPLE  SEVERAL  WOODS  
2   BROWN  SEVERAL  WOODS  
3  PURPLE  SEVERAL  WOODS  
4   BROWN  SEVERAL  WOODS  

[5 rows x 23 columns]

In [12]:
data_characteristics = pd.read_csv("agaricus-lepiota.data")

In [13]:
data_characteristics.head()

p  x  s  n  t p.1  f  c n.1  k  ... s.2  w w.1 p.2 w.2  o p.3 k.1 s.3  u
0  e  x  s  y  t   a  f  c   b  k  ...   s  w   w   p   w  o   p   n   n  g
1  e  b  s  w  t   l  f  c   b  n  ...   s  w   w   p   w  o   p   n   n  m
2  p  x  y  w  t   p  f  c   n  n  ...   s  w   w   p   w  o   p   k   s  u
3  e  x  s  g  f   n  f  w   b  k  ...   s  w   w   p   w  o   e   n   a  g
4  e  x  y  y  t   a  f  c   b  n  ...   s  w   w   p   w  o   p   k   n  g

[5 rows x 23 columns]

## Understanding the data

This data set includes descriptions of hypothetical samples corresponding to 23 species of gilled mushrooms in the Agaricus and Lepiota Family (pp. 500-525).  Each species is identified as definitely edible, definitely poisonous, or of unknown edibility and not recommended.  This latter class was combined with the poisonous one.  The Guide clearly states that there is no simple rule for determining the edibility of a mushroom; no rule like ``leaflets three, let it be'' for Poisonous Oak and Ivy.

1. Adding titles to organize the info

In [14]:
classes = np.array([
    'class',
    'cap-shape', 
    'cap-surface', 
    'cap-color',
    'bruises?',
    'odor',
    'gill-attachment',
    'gill-spacing',
    'gill-size',
    'gill-color',
    'stalk-shape',
    'stalk-root',
    'stalk-surface-above-ring',
    'stalk-surface-below-ring',
    'stalk-color-above-ring',
    'stalk-color-below-ring',
    'veil-type',
    'veil-color',
    'ring-number',
    'ring-type',
    'spore-print-color',
    'population',
    'habitat'])

featured_classes = np.array([
    'cap-shape', 
    'cap-surface', 
    'cap-color',
    'bruises?',
    'odor',
    'gill-attachment',
    'gill-spacing',
    'gill-size',
    'gill-color',
    'stalk-shape',
    'stalk-root',
    'stalk-surface-above-ring',
    'stalk-surface-below-ring',
    'stalk-color-above-ring',
    'stalk-color-below-ring',
    'veil-type',
    'veil-color',
    'ring-number',
    'ring-type',
    'spore-print-color',
    'population',
    'habitat'])

In [23]:
dataset_names = pd.DataFrame(data_names.values, columns=classes)
X, y = pd.DataFrame(dataset_names, columns=featured_classes), pd.DataFrame(dataset_names, columns=['class'])


In [29]:
numData = 80
X_enc = MultiColumnLabelEncoder().fit_transform(X.head(numData))


In [30]:
# Transferring to numerical values using LabelEncoder

#X_enc = pd.DataFrame(dataset_encoded, columns=featured_classes)
y_enc = MultiColumnLabelEncoder().fit_transform(y.head(numData)) #pd.DataFrame(dataset_encoded, columns=['class'])

In [31]:
dataset_names.head()

class cap-shape cap-surface cap-color bruises?    odor gill-attachment  \
0  EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND            FREE   
1  EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND            FREE   
2  EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND            FREE   
3  EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND            FREE   
4  EDIBLE    CONVEX      SMOOTH     WHITE  BRUISES  ALMOND            FREE   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0      CROWDED    NARROW      WHITE  ...                   SMOOTH   
1      CROWDED    NARROW       PINK  ...                   SMOOTH   
2      CROWDED    NARROW       PINK  ...                   SMOOTH   
3      CROWDED    NARROW      BROWN  ...                   SMOOTH   
4      CROWDED    NARROW      BROWN  ...                   SMOOTH   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                  WHITE                  WHITE   PARTIAL      WHITE   
1                  WHITE                  WHITE   PARTIAL      WHITE   
2                  WHITE                  WHITE   PARTIAL      WHITE   
3                  WHITE                  WHITE   PARTIAL      WHITE   
4                  WHITE                  WHITE   PARTIAL      WHITE   

  ring-number ring-type spore-print-color population habitat  
0         ONE   PENDANT             BROWN    SEVERAL   WOODS  
1         ONE   PENDANT            PURPLE    SEVERAL   WOODS  
2         ONE   PENDANT             BROWN    SEVERAL   WOODS  
3         ONE   PENDANT            PURPLE    SEVERAL   WOODS  
4         ONE   PENDANT             BROWN    SEVERAL   WOODS  

[5 rows x 23 columns]

In [32]:
X_enc.head()

cap-shape  \
0  [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                         cap-surface  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                           cap-color  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                            bruises?  \
0  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                                odor  \
0  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                     gill-attachment  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, ...   

                                        gill-spacing  \
0  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                           gill-size  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, ...   

                                          gill-color  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, ...   
3  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                                         stalk-shape  ...  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  ...   

                            stalk-surface-below-ring  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   

                              stalk-color-above-ring  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
1  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
2  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
3  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   
4  [0, 0, 0, 0, 0, 0, 0, 0

### Possible approach

We can translate the classes to numerical data. This way we could try linear regression methods.
> Of course it may make no sense, if the outcome is 'between' labels, but it can yield sense in a way, as some classes are more fluid than others (i.E. colors: A mushroom can be in between yellow and white (beeing bright yellow))
>
>Even though we have no data on this level of accuracy, we might build a regression model that supports it.

Also we could use correlation methods i.e. Pearsons Correlation to identify correlations between classes, this way we might find correlations that are not visible or more than 2-Dimensional.


In [33]:
## How do we define a good value for n_neighbors or k value?
# Using a square root of the data length and checking that it is an odd number!

#import math
#math.sqrt(len(y_test))
## 82.04876598706406

## So, the optimal value for k is 82 - 1
n_neighbors = 81

In [34]:
# PCA
y_labels = y_enc.astype(int).values[:,-1]
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.neighbors import (KNeighborsClassifier,
                               NeighborhoodComponentsAnalysis)
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

random_state = 42

# Split into train/test
X_train, X_test, y_train, y_test = train_test_split(X_enc, y_enc, test_size=0.8, stratify=y_labels,
                     random_state=random_state)

# Reduce dimension to 2 with PCA
pca = make_pipeline(StandardScaler(),
                    PCA(n_components=2, random_state=random_state))

# Use a nearest neighbor classifier to evaluate the methods
# p = 2 we are looking for Mushroom edible or not
knn = KNeighborsClassifier(n_neighbors=n_neighbors,p=2,metric='euclidean')
plt.figure()
pca.fit(X_train, y_train)

# Fit a nearest neighbor classifier on the embedded training set
knn.fit(pca.transform(X_train), y_train)

# Compute the nearest neighbor accuracy on the embedded test set
acc_knn = knn.score(pca.transform(X_test), y_test)

# Embed the data set in 2 dimensions using the fitted model
X_embedded = pca.transform(X_enc)

# Plot the projected points and show the evaluation score
plt.scatter(X_embedded[:, 0], X_embedded[:, 1], c=y_labels, s=10)
plt.title("{}, KNN (k={})\nTest accuracy = {:.2f}".format('PCA', n_neighbors, acc_knn))
plt.show()

ValueError: setting an array element with a sequence.

<Figure size 432x288 with 0 Axes>

This looks like knn gives us a good fitting method for testing additional samples. 
The benefit is that knn works also on smaller dimensions. (IE. if we only have 3-4 dimensions of information from a new mushroom, that we want to classify now)




### Evaluating the model

In [29]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

In [30]:
y_pred = knn.predict(pca.transform(X_test))

print(len(y_pred), y_pred.shape,len(y_test), len(X_test), X_test.shape)
print(y_pred)

6732 (6732,) 6732 6732 (6732, 22)
[0 0 1 ... 1 0 0]


In [31]:
# Using confusion matrix to evaluate the y predicted data
cm = confusion_matrix(y_test,y_pred)

<img src="./resources/confusion_table_ef.JPG" width="" />

In [71]:
cm

array([[3087,  503],
       [ 109, 3033]], dtype=int64)

- As result we have 3087 mushroom edible species.
- 3033 poisonous mushrooms.
- 109 possible errors
- 503 false positives. 

In [81]:
f1_score(y_test,y_pred)

0.9083557951482479

In [82]:
accuracy_score(y_test,y_pred)

0.9090909090909091

## Reference

(1) Metzler S. Metzler V. Orson K. Texas Mushrooms. University of Texas Press. Austin. https://books.google.cz/books?id=HRtfvVigMmsC&pg=PA37&redir_esc=y#v=onepage&q&f=false <br>
(2) Haytowitz. Vitamin D in Mushrooms. Nutrient Data Laboratory, Beltsville Human Nutrition Research Center. <br>
(3) Indiana State Parks - Common Mushrooms <br>
(4) British Fungi - Elementary Text-Book <br>
(5) https://www.usask.ca/biology/fungi/home_%20page.shtml